### PR TITLE
Spark: Spark SQL Extensions for create tag

### PR DESCRIPTION
**Pull request by hililiwei**
Co-authored-by: Amogh Jahagirdar <jahamogh@amazon.com>
Co-authored-by: chidayong <247070443@qq.com>

## What is the purpose of the change
Implement the syntax in the following documents:
https://docs.google.com/document/d/1tbATFPrKF3vNlzkgZQdaW8CAJmbjvryfrlg6C2Ci_aA/edit

### Create TAG 

```sql
ALTER TABLE table CREAT TAG tagName
[AS OF {VERSION snapshotId}]
[RETAIN interval {DAYS | HOURS | MINUTES}]
```
e.g.
```
ALTER TABLE table CREATE TAG historical_annual_snapshot
AS OF VERSION 100
RETAIN 12 DAYS
```

cc @jackye1995 @amogh-jahagirdar

### CODE REVIEW

**Comment by hililiwei** [comment]
In the original proposal, it was "[RETAIN For interval {DAYS | HOURS | MINUTES}]", but in keeping with CREATE BRANCH, I removed the" For" key.  What do you think about that? @jackye1995 @amogh-jahagirdar

**Comment by jackye1995** [comment]
Good point, +1 for only RETAIN because https://docs.databricks.com/sql/language-manual/delta-vacuum.html

**Comment by amogh-jahagirdar** [comment]
> In the original proposal, it was "[RETAIN For interval {DAYS | HOURS | MINUTES}]", but in keeping with CREATE BRANCH, I removed the" For" key. What do you think about that? @jackye1995 @amogh-jahagirdar

Ah yes, we should remove the FOR clause. That's a mistake in the proposal on my part. Updated the proposal.

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 1]
```diff
@@ -73,6 +73,7 @@ statement
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
     | ALTER TABLE multipartIdentifier SET IDENTIFIER_KW FIELDS fieldList                    #setIdentifierFields
     | ALTER TABLE multipartIdentifier DROP IDENTIFIER_KW FIELDS fieldList                   #dropIdentifierFields
+    | ALTER TABLE multipartIdentifier CREATE TAG identifier (AS OF VERSION snapshotId)?  (RETAIN snapshotRefRetain snapshotRefRetainTimeUnit)? #createTag
```
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: I am debating with myself, if we should merge this case with the CREATE BRANCH case or not. 

I guess this question will also come for REPLACE BRANCH.

Currently leaning towards keep it as is so it is more clear, each extension just does 1 exact thing, similar to we have 2 different actions for SET/DROP identifier field, 3 for ADD/DROP/REPLACE partition field. 

Any thoughts? @amogh-jahagirdar @hililiwei

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 2]
```diff
@@ -168,34 +169,61 @@ fieldList
...
+    | DISTRIBUTED | LOCALLY | MINUTES | UNORDERED | REPLACE | VERSION | WITH | IDENTIFIER_KW | FIELDS | SET
+    | TAG | TRUE | FALSE
     | MAP
     ;
 
+snapshotId
+    : number
+    ;
+
+snapshotRefRetain
```
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:182]: I overlooked this in the last PR, do we really need this extra definition? 

To me, it feels more intuitive to just say `RETAIN number timeUnit` instead of `RETAIN snapshotRefRetain snapshotRefRetainTimeUnit`

Another thing I overlooked is that looks like the Antlr convention is to use all capital letters for these definitions, like `TIME_UNIT` instead of `timeUnit`. I don't know if they imply different functionalities though.

**Review by jackye1995 — COMMENTED** [review]
Overall looks good to me, I will ignore the duplicated code and do another pass after rebase. Added 2 comments for discussion.

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 (hunk 1 above)
Comment (amogh-jahagirdar) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4]: I think it make sense just to keep it separate as well. It's nice just to read the sql extension and immediately see that CREATE TAG only supports RETAIN (there's no notion of min snapshots to keep). Same principle applies with CREATE BRANCH (can see what are the clauses that can be applied). Otherwise one has to read the actual action logic to verify that.

Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4]: Agree with @amogh-jahagirdar.

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 (hunk 2 above)
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4]: >To me, it feels more intuitive to just say RETAIN number timeUnit instead of RETAIN snapshotRefRetain snapshotRefRetainTimeUnit

I originally added `snapshotRefRetain` and `snapshotRetain` to make the statement parsing code more readable. Removing it is technically feasible. In the new version, I have removed (including create branch).

>Another thing I overlooked is that looks like the Antlr convention is to use all capital letters for these definitions, like TIME_UNIT instead of timeUnit. I don't know if they imply different functionalities though.


All caps usually indicate keywords. So, I prefer keep `timeUnit`.
```
ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
```

like `TABLE`:
`TABLE: 'TABLE';`

However, `writeSpec` is written as:
```
writeSpec
    : (writeDistributionSpec | writeOrderingSpec)*
    ;
```
`writeDistributionSpec`:
```
writeDistributionSpec
    : DISTRIBUTED BY PARTITION
    ;
```

Thanks.
Liwei

**Comment by amogh-jahagirdar** [comment]
@hililiwei similar to https://github.com/apache/iceberg/pull/6638 could this PR encapsulate create/replace? We came to the conclusion on the replace PR it made more sense to just combine them. at the end of the day the logical plan needs to pass through whether it's a replace or not and it is a bit more concise in the grammar definition to combine them while still maintaining readability

**Comment by jackye1995** [comment]
> could this PR encapsulate create/replace?

+1

Let me know when this is updated, I will take another look!

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 3]
```diff
@@ -73,14 +73,22 @@ statement
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
     | ALTER TABLE multipartIdentifier SET IDENTIFIER_KW FIELDS fieldList                    #setIdentifierFields
     | ALTER TABLE multipartIdentifier DROP IDENTIFIER_KW FIELDS fieldList                   #dropIdentifierFields
-    | ALTER TABLE multipartIdentifier createReplaceBranchClause   #createOrReplaceBranch
+    | ALTER TABLE multipartIdentifier createReplaceBranchClause                             #createOrReplaceBranch
+    | ALTER TABLE multipartIdentifier createTagClause                                       #createTag
+    ;
+
+createTagClause
+    : CREATE TAG (IF NOT EXISTS)? identifier tagOptions
```
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:81]: not sure if you finished all updates or not, but I think we can also add REPLACE part here and make it `#createorReplaceTag`?

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 4]
```diff
@@ -73,14 +73,22 @@ statement
...
     ;
 
 createReplaceBranchClause
     : (CREATE OR)? REPLACE BRANCH identifier branchOptions
     | CREATE BRANCH (IF NOT EXISTS)? identifier branchOptions
     ;
 
+tagOptions
+    : (AS OF VERSION snapshotId)? (refRetain)?;
```
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:90]: hmm I just noticed this `;` in the end, also in `branchOptions`. @amogh-jahagirdar is this required? It doesn't feel like required to me.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala [hunk 5]
```diff
@@ -206,8 +206,9 @@ class IcebergSparkSqlExtensionsParser(delegate: ParserInterface) extends ParserI
             normalized.contains("write unordered") ||
             normalized.contains("set identifier fields") ||
             normalized.contains("drop identifier fields") ||
-            normalized.contains("create branch"))) ||
-            normalized.contains("replace branch")
+            normalized.contains("create branch") ||
+            normalized.contains("replace branch") ||
+            normalized.contains("create tag")))
```
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala:211]: I think there is a fix for this at #6728 , might want to rebase after that is merged

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala [hunk 6]
```diff
@@ -92,14 +94,17 @@ class IcebergSqlExtensionsAstBuilder(delegate: ParserInterface) extends IcebergS
       typedVisit[Transform](ctx.transform))
   }
 
+  /**
+   * Create an CREATE OR REPLACE BRANCH logical command.
+   */
```
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:99]: nice catches! But can we avoid unrelated changes in the PR? @amogh-jahagirdar I think you should fix these separately

Comment (jackye1995) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:99]: btw I am surprised that checkstyle did not catch that extra space, maybe it needs some more configuration.

Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:99]: For scala, our checkstyle needs to be improved. And auto-formatting doesn't work yet.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala (hunk 5 above)
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala]: got it.

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 (hunk 4 above)
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:91]: this is required. We can insert a new line, consistent with the rest.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala (hunk 6 above)
Comment (amogh-jahagirdar) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:99]: Great catch @hililiwei , I'll update #6728 to fix the styling issues. We should track an issue for scala checkstyle/autoformatting , admittedly I rely on that quite heavily to catch these kinds of issues.

Comment (amogh-jahagirdar) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:99]: Created https://github.com/apache/iceberg/issues/6736 and updated #6728 to include styling fixes

Comment (jackye1995) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:99]: Thanks, I merged #6728

**Comment by jackye1995** [comment]
Apart from the style changes for branch that can now be removed, I have no further comments. Let me know when this is rebased!

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala [hunk 7]
```diff
@@ -128,6 +133,36 @@ class IcebergSqlExtensionsAstBuilder(delegate: ParserInterface) extends IcebergS
 
   }
 
+
```
Comment (aokolnychyi) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:136]: nit: extra line

**Comment by aokolnychyi** [comment]
cc @RussellSpitzer @szehon-ho @flyrain @karuppayya

**Comment by hililiwei** [comment]
> hililiwei requested review from jackye1995 and removed request for flyrain and yyanyy

Misclick. 😂

> Apart from the style changes for branch that can now be removed, I have no further comments. Let me know when this is rebased!

PTAL, thx.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala (hunk 7 above)
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala]: fixed.

**Review by jackye1995 — APPROVED** [review]
Mostly looks good to me! @amogh-jahagirdar @aokolnychyi any further comments?

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestReplaceTag.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+import org.apache.iceberg.spark.source.SimpleRecord;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runners.Parameterized;
+
+public class TestReplaceTag extends SparkExtensionsTestBase {
```
Comment (amogh-jahagirdar) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestReplaceTag.java:38]: Can we move all the tag tests to TestTagDDL based on the discussion here https://github.com/apache/iceberg/pull/6752#discussion_r1097916794? Not blocking for this PR we can get in a follow on, but it's basically just to ensure the test runs aren't super slow with all the spark setup which will happen for every test class if we have a test class per statement.

**Review by amogh-jahagirdar — APPROVED** [review]
Thanks LiWei, one non-blocking comment on our test setup for all the branching/tagging DDLs. We can address on a follow on PR.

**Review by yyanyy — APPROVED** [review]
lgtm!

**Comment by jackye1995** [comment]
All the comments are addressed and we have enough votes, so I will go ahead to merge it. I requested @aokolnychyi for further review and approval, but given it has been quite a few days, and there is a subsequent DROP TAG PR anyway, we can move forward and address any missing comments there.

Thanks for the contribution @hililiwei , and thanks everyone for the review!
